src/main/java/com/hivemq/persistence/local/memory/ClientSessionMemoryLocalPersistence.java

```diff
@@ -0,0 +1,531 @@
+import com.hivemq.persistence.NoSessionException;
+import com.hivemq.persistence.PersistenceEntry;
+import com.hivemq.persistence.PersistenceFilter;
+import com.hivemq.persistence.clientsession.ClientSession;
+import com.hivemq.persistence.clientsession.ClientSessionWill;
+import com.hivemq.persistence.clientsession.PendingWillMessages;
+import com.hivemq.persistence.exception.InvalidSessionExpiryIntervalException;
+import com.hivemq.persistence.local.ClientSessionLocalPersistence;
+import com.hivemq.persistence.local.xodus.BucketChunkResult;
+import com.hivemq.persistence.local.xodus.bucket.BucketUtils;
+import com.hivemq.persistence.payload.PublishPayloadPersistence;
+import com.hivemq.util.ClientSessions;
+import com.hivemq.util.ThreadPreConditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.stream.Collectors;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.hivemq.mqtt.message.connect.Mqtt5CONNECT.SESSION_EXPIRE_ON_DISCONNECT;
+import static com.hivemq.mqtt.message.connect.Mqtt5CONNECT.SESSION_EXPIRY_NOT_SET;
+import static com.hivemq.util.ThreadPreConditions.SINGLE_WRITER_THREAD_PREFIX;
+
+/**
+ * @author Georg Held
+ */
+public class ClientSessionMemoryLocalPersistence implements ClientSessionLocalPersistence {
+
+    private static final @NotNull Logger log = LoggerFactory.getLogger(ClientSessionMemoryLocalPersistence.class);
+
+    final @NotNull AtomicLong currentMemorySize = new AtomicLong();
+
+    private final @NotNull PublishPayloadPersistence payloadPersistence;
+    private final @NotNull EventLog eventLog;
+
+    private final long configuredSessionExpiryInterval;
+    private final @NotNull AtomicInteger sessionsCount = new AtomicInteger(0);
+
+    private final int bucketCount;
+    private final @NotNull Map<String, PersistenceEntry<ClientSession>>[] buckets;
+
+    @Inject
+    ClientSessionMemoryLocalPersistence(
+            final @NotNull MqttConfigurationService mqttConfigurationService,
+            final @NotNull PublishPayloadPersistence payloadPersistence,
+            final @NotNull MetricRegistry metricRegistry,
+            final @NotNull EventLog eventLog) {
+
+        this.payloadPersistence = payloadPersistence;
+        this.eventLog = eventLog;
+        this.configuredSessionExpiryInterval = mqttConfigurationService.maxSessionExpiryInterval();
+
+        bucketCount = InternalConfigurations.PERSISTENCE_BUCKET_COUNT.get();
+
+        //noinspection unchecked
+        buckets = new ConcurrentHashMap[bucketCount];
+        for (int i = 0; i < bucketCount; i++) {
+            buckets[i] = new ConcurrentHashMap<>();
```

Comment:
Why concurrent? please use a normal HashMap, that is the reason why we have buckets